File: app/src/main/java/com/printf/kidsteacher/fragment/VideoFragment.java
package com.printf.kidsteacher.fragment;

import android.app.Activity;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.ActivityInfo;
import android.os.Bundle;
import android.os.Handler;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.ImageView;
import android.widget.LinearLayout;

import com.google.gson.Gson;
import com.printf.kidsteacher.R;
import com.printf.kidsteacher.adapter.VideoAdapter;
import com.printf.kidsteacher.been.videoData.VideoData;
import com.printf.kidsteacher.common.ApiCall;
import com.printf.kidsteacher.common.ApiResponce;
import com.printf.kidsteacher.common.CheckInternet;
import com.printf.kidsteacher.common.WebServices;
import com.printf.kidsteacher.view.CustomTextView;

public class VideoFragment extends BaseFragment {
    private BroadcastReceiver mReceiver;
    RecyclerView rv_video;
    VideoAdapter videoAdapter;
    VideoData videoData;
    LinearLayout ll_noInternet;
    ImageView iv_refresh;
    Animation animation;
    CustomTextView tv_message;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return  inflater.inflate(R.layout.fragment_video, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        init(view);
    }

    private void init(View view) {
        rv_video = view.findViewById(R.id.rv_video);
        ll_noInternet = view.findViewById(R.id.ll_noInternet);
        tv_message = view.findViewById(R.id.tv_message);
        iv_refresh = view.findViewById(R.id.iv_refresh);
        callApi();
        //new Handler().postDelayed(new Runnable() {@Override public void run() { callApi(); }},1000);

        animation = AnimationUtils.loadAnimation(getActivity(), R.anim.bounce);
        animation.setAnimationListener(new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(Animation animation) {

            }

            @Override
            public void onAnimationEnd(Animation animation) {
               callApi();
            }

            @Override
            public void onAnimationRepeat(Animation animation) {

            }
        });
        iv_refresh.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                iv_refresh.startAnimation(animation);
            }
        });
    }

    private void callApi() {
        if (!CheckInternet.networkAvailability(activity)) {
            tv_message.setText("No internet connection. Please connect to internet and try again.");
            ll_noInternet.setVisibility(View.VISIBLE);
            return;
        }
        ApiCall.GetApi(true, false, activity, WebServices.VIDEO_API, new ApiResponce() {
            @Override
            public void Responce(String responce) {

                if(isVisible() && isAdded()){
                    ll_noInternet.setVisibility(View.GONE);
                    videoData = new Gson().fromJson(responce, VideoData.class);
                    setAdapter(videoData);
                    if (videoData.getData().size() == 0) {
                        tv_message.setText("Awesome videos coming soon. We are creating awesome videos for you.");
                        ll_noInternet.setVisibility(View.VISIBLE);
                    }
                }
            }

            @Override
            public void Error(String error) {

            }
        });
    }

    private void setAdapter(VideoData videoData) {
        GridLayoutManager gridLayoutManager = new GridLayoutManager(getActivity(), 3);
        rv_video.setLayoutManager(gridLayoutManager);
        videoAdapter = new VideoAdapter(getActivity(), videoData.getData());
        rv_video.setAdapter(videoAdapter);
    }

    @Override
    public void onResume() {
        super.onResume();

        IntentFilter intentFilter = new IntentFilter("android.intent.action.MAIN");
        mReceiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                String tempSearchedText = intent.getStringExtra("Search");

                if (videoAdapter == null) {
                    return;
                }
                videoAdapter.filter(tempSearchedText);
            }
        };
        //registering our receiver
        getActivity().registerReceiver(mReceiver, intentFilter);

    }

       @Override
    public void setUserVisibleHint(boolean isVisibleToUser) {
        super.setUserVisibleHint(isVisibleToUser);
        if (isVisibleToUser) {
            Activity a = getActivity();
            if (a != null) a.setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
        }
    }
}

File: app/src/main/java/com/printf/kidsteacher/fragment/VideoFragment.kt
package com.printf.kidsteacher.fragment

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import androidx.lifecycle.LifecycleObserver
import androidx.lifecycle.Observer
import androidx.lifecycle.ProcessLifecycleOwner
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.GridLayoutManager
import com.printf.kidsteacher.R
import com.printf.kidsteacher.activity.DetailViewModel
import com.printf.kidsteacher.activity.VideoDetailActivity
import com.printf.kidsteacher.adapter.VideoAdapter
import com.printf.kidsteacher.been.videoData.VideoData
import com.printf.kidsteacher.common.CheckInternet
import com.printf.kidsteacher.other.RecyclerViewClick
import com.printf.kidsteacher.video.VideoViewModel
import com.roger.catloadinglibrary.CatLoadingView
import kotlinx.android.synthetic.main.fragment_video.*
import java.io.Serializable

class VideoFragment : BaseFragment(), LifecycleObserver, RecyclerViewClick {

    private lateinit var viewModel: DetailViewModel
    var videoAdapter: VideoAdapter? = null
    var animation: Animation? = null

    private lateinit var videoViewModel: VideoViewModel

    private lateinit var mView: CatLoadingView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewModel = activity?.run {
            ViewModelProviders.of(this)[DetailViewModel::class.java]
        } ?: throw Exception("Invalid Activity")

        viewModel.searchObservable.observe(this, Observer<String> { search ->
            if(videoAdapter != null)
                videoAdapter!!.filter(search)
        })

        mView = CatLoadingView()
        mView.setCanceledOnTouchOutside(false)

    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_video, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        videoViewModel = VideoViewModel(requireActivity().application)
        setupObservers()

        init()
    }

    private fun init() {
        callApi()
        animation = AnimationUtils.loadAnimation(activity, R.anim.bounce)
        animation?.setAnimationListener(object : Animation.AnimationListener {
            override fun onAnimationStart(animation: Animation) {}
            override fun onAnimationEnd(animation: Animation) {
                callApi()
            }

            override fun onAnimationRepeat(animation: Animation) {}
        })
        iv_refresh.setOnClickListener { iv_refresh?.startAnimation(animation) }
    }

    private fun callApi() {
        if (!CheckInternet.networkAvailability(requireContext())) {
            ll_noInternet.visibility = View.VISIBLE
            rv_video.visibility = View.GONE
            return
        }
        videoViewModel.getVideo()
    }

    private fun setAdapter(videoData: VideoData?) {
        val gridLayoutManager = GridLayoutManager(requireContext(), 3)
        rv_video!!.layoutManager = gridLayoutManager
        videoAdapter = VideoAdapter(requireContext(), videoData!!.data!!, this)
        rv_video!!.adapter = videoAdapter
    }

    private fun setupObservers() {
        videoViewModel.dataObservable.observe(viewLifecycleOwner, Observer {
            ll_noInternet!!.visibility = View.GONE
            if (it.data.isNullOrEmpty() || it.data?.size == 0) {
                ll_noInternet.visibility = View.VISIBLE
                rv_video.visibility = View.GONE
            } else {
                ll_noInternet.visibility = View.GONE
                rv_video.visibility = View.VISIBLE
                setAdapter(it)
            }
        })
        videoViewModel.isLoading.observe(viewLifecycleOwner, Observer {
            if (it) {
                mView.show(requireActivity().supportFragmentManager, "")
            } else {
                mView.dismiss()
            }
        })

        videoViewModel.errorObservable.observe(viewLifecycleOwner, Observer {
            ll_noInternet.visibility = View.VISIBLE
            rv_video.visibility = View.GONE
        })

        ProcessLifecycleOwner.get().lifecycle.removeObserver(this)
        ProcessLifecycleOwner.get().lifecycle.addObserver(this)
    }

    override fun OnClick(img: Int, name: String?, position: Int) {
        val intent = Intent(context, VideoDetailActivity::class.java)
        val bundle = Bundle()
        bundle.putSerializable("list", videoViewModel.dataObservable.value?.data!! as Serializable)
        bundle.putInt("position", position)
        intent.putExtra("list", bundle)

        viewModel.setIntent(intent)
    }
}